### PR TITLE
[TIMOB-20522] Use FusedLocationProviderClient for Ti.Geolocation

### PR DESCRIPTION
**Pull request by garymathews**
- Use `FusedLocationProviderClient` when `ti.playservices` module is included in `tiapp.xml`, falling back to previous behavior if not available

###### TEST CASE
```JS
var win = Ti.UI.createWindow({ backgroundColor: 'gray' }),
    listView = Ti.UI.createListView(),
    count = 1;

Ti.Geolocation.accuracy = Ti.Geolocation.ACCURACY_HIGH;

function push (e) {
    var section = Ti.UI.createListSection({ headerTitle: '#' + count++ });
    if (e.success) {
        if (e.coords) {
            e = e.coords;
        }
        section.setItems([
            { properties: { title: 'LOCATION:\n' + e.latitude + ', ' + e.longitude, color: 'orange' } },
            { properties: { title: 'ALTITUDE:\n' + e.altitude, color: 'orange' } },
            { properties: { title: 'ACCURACY:\n' + e.accuracy, color: 'orange' } }
        ]);
    } else {
        section.setItems([
            { properties: { title: 'ERROR:\n' + e.error, color: 'red' } }
        ]);
    }
    listView.appendSection(section);
}

function getLocation() {
    Ti.Geolocation.addEventListener('location', push);
    Ti.Geolocation.getCurrentPosition(push);
}

win.addEventListener('open', function(){
    if(Ti.Geolocation.hasLocationPermissions()){
        getLocation();
    }else{
        Ti.Geolocation.requestLocationPermissions(Ti.Geolocation.AUTHORIZATION_ALWAYS, function (e) {
            if (e.success) {
                getLocation();
            } else {
                alert('could not obtain location permissions');
            }
        });
    }
});

win.add(listView);
win.open();
```

NOTE: The play services libraries for the Geolocation module are only required to compile the SDK. They are not included with titanium builds and won't cause conflicts with `ti.playservices`. However, they will need updating if the Google API changes.

[JIRA Ticket](https://jira.appcelerator.org/browse/TIMOB-20522)

### CODE REVIEW

**Comment by hansemannn** [comment]
@garymathews Does this also affect https://github.com/appcelerator/titanium_mobile/pull/9633?

**Comment by garymathews** [comment]
@hansemannn https://github.com/appcelerator/titanium_mobile/pull/9633 is not affected by this

File: apidoc/Titanium/Geolocation/Geolocation.yml [hunk 1]
```diff
@@ -395,6 +428,8 @@ properties:
     type: Number
     permission: read-only
     platforms: [android, iphone, ipad]
+    deprecated:
+        since: "2.0.0"
```
Comment (hansemannn) [inline apidoc/Titanium/Geolocation/Geolocation.yml]: I don't think they are deprecated for iOS. Maybe only deprecate and remove for Android?

**Comment by lokeshchdhry** [comment]
FR Passed.
`FusedLocationProviderClient` works as expected when `ti.playservices` module is included in the project & falls back to old behaviour if not.

Also, No crash is seen if wrong accuracy  like `Ti.Geolocation.ACCURACY_BEST_FOR_NAVIGATION` is provided which is not meant for Android.

Studio Ver:     5.0.0.201712081732
SDK Ver:        7.2.0 local build
OS Ver:         10.13.2
Xcode Ver:      Xcode 9.2
Appc NPM:       4.2.12
Appc CLI:       7.0.2
Daemon Ver:     1.0.1
Ti CLI Ver:     5.0.14
Alloy Ver:      1.11.0
Node Ver:       8.9.1
NPM Ver:        5.5.1
Java Ver:       1.8.0_101
Devices:        ⇨ google Nexus 6P --- Android 8.0.0
                ⇨ google Nexus 5 --- Android 6.0.1

**Comment by jquick-axway** [comment]
@lokeshchdhry, can you also try testing this with the `ti.playservices` module included on an Android emulator that does **NOT** have Google Play on it please?

The reason is Google Play Services is an APK that is installed on the device. It's a background process that the `ti.playservices` communicates with to acquire location data from. So, it's a 3rd test case. In this case, it should still fallback to using the old behavior.

**Comment by lokeshchdhry** [comment]
@jquick-axway , Sure thanks for the suggestion.

**Comment by lokeshchdhry** [comment]
@garymathews , On a genymotion emulator with no playservices I don't see any location events. I think its not falling back to old behaviour.

File: android/modules/geolocation/src/java/ti/modules/titanium/geolocation/android/FusedLocationProvider.java [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+	 * @return do we have access to Google Play Services APIs?
+	 */
+	public static boolean hasPlayServices()
+	{
+		try {
+			Class.forName("com.google.android.gms.common.GoogleApiAvailability");
+		} catch (ClassNotFoundException e) {
+			return false;
+		}
+		return PlayServices.validVersion();
```
Comment (jquick-axway) [inline android/modules/geolocation/src/java/ti/modules/titanium/geolocation/android/FusedLocationProvider.java]: We should also check if `isGooglePlayServicesAvailable()` returns `SUCCESS` in the `hasPlayServices()` method.

This is to handle the case where `ti.playservices` module is included, but we're running on a device/emulator that doesn't have Google Play Services installed, is disabled, or is too old.

**Review by jquick-axway — CHANGES_REQUESTED** [review]

**Comment by lokeshchdhry** [comment]
Oops forgot to comment before merging.

Checked the changes & geolocation works as expected in emulator without google playservices.

Studio Ver: 5.0.0.201712081732
SDK Ver: 7.2.0 local build
OS Ver: 10.13.2
Xcode Ver: Xcode 9.2
Appc NPM: 4.2.12
Appc CLI: 7.0.2
Daemon Ver: 1.0.1
Ti CLI Ver: 5.0.14
Alloy Ver: 1.11.0
Node Ver: 8.9.1
NPM Ver: 5.5.1
Java Ver: 1.8.0_101
Devices: ⇨ google Nexus 6P --- Android 8.0.0
⇨ google Nexus 5 --- Android 6.0.1

**Pull request merged**